[PGNCCL] Stash tensors for reduce_scatter_v and all_gather_v

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #148994

https://github.com/pytorch/pytorch/pull/148590 removed `record_stream`. Since previous `AVOID_RECORD` flag does not cover `reduce_scatter_v` and `all_gather_v` which are in coalescing form, these two ops were missed. Causing TorchRec's Variable Length Embedding to fail.

This PR adds a vector to stash tensors when coalescing is in flight. And the end of coalescing, it will hand over the tensors to `Work`. 

Rest of the PR is mostly BE -- grouping various variables related to coalescing in one single struct and offer a `reset` method. So that it is easier to extend what we need to temporarily bookkeep.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

Differential Revision: [D71589949](https://our.internmc.facebook.com/intern/diff/D71589949)

kwen2501 committed:
[PGNCCL] Group coalescing states in one struct

[ghstack-poisoned]

eqy commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -235,6 +235,28 @@ struct DumpPipe {
+  // Whether the coalesced calls are sync or async.
+  bool async = false;
+  // Stores the list of tensors for all collectives run inside a coalescing
+  // TODO (eqy)
```
One piece of context I'm missing is how exactly the operations diverge between CoalColl and CoalP2P. From some crude instrumentation of the code I observe that `pointToPoint` is called in `batch_isend_irecv` but there is no call to something like `pointToPointCoalesced` (that doesn't exist AFAIK), whereas I'm guessing `collectiveCoalesced` is called in the `CoalColl` case. The interesting thing is that `work` is not initialized in `pointToPoint` when `coalescingState_` is `true`, but it seems to be initialized in `collectiveCoalesced` even when `coalescingState_` is `true`, and this is how `avoidRecordStreams_` is handled for the `collectiveCoalesced` case.

I'm mainly curious where this difference comes from, and would the consequence of that be that we would really only stash tensors here for the `CoalP2P` case?

kwen2501 replied:
Oh ignore `collectiveCoalesced`. It is unrelated here. 
`collectiveCoalesced` is a shared helper for `allreduce_coalesced`, `allgather_into_tensor_coalesced` and `reduce_scatter_tensor_coalesced`, calling collectives on the vector of tensors in a loop (NCCL aggregation).

It is almost orthogonal to `startCoalescing` and `endCoalescing` which mainly interact with `pointToPoint` as you mentioned.

fduwjj replied:
Do we document this somewhere? Sounds like tribal knowledge to me..

kwen2501 replied:
I think I will have a PR to remove `collectiveCoalesced` entirely.

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -235,6 +235,28 @@ struct DumpPipe {
 };
 #endif
 
+static constexpr int CoalColl = 0x02, CoalP2P = 0x04;
```
why not int CoalColl = 0x01, CoalP2P = 0x02;?

kwen2501 replied:
Just staying the same as previous values.

fduwjj replied:
but since we control it and it is only used within PGNCCL. does it make sense to start with 0x01?

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -1221,16 +1243,7 @@ class TORCH_API ProcessGroupNCCL : public Backend {
-
-  // Whether the coalesced calls are sync or async.
-  bool coalescedAsync_;
+  struct CoalescingState coalescingState_;
```
Do you think if we call it `coalescingMode_` would be better? No strong opinion.

kwen2501 committed:
Update on "[PGNCCL] Group coalescing states in one struct"


Mostly BE. Grouping various variables related to coalescing in one single struct and offer a `reset` method. So that it is easier to extend the states we need to temporarily bookkeep when coalescing is in flight.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
@kwen2501 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71589949).

kwen2501 committed:
Update on "[PGNCCL] Group coalescing states in one struct"


Mostly BE. Grouping various variables related to coalescing in one single struct and offer a `reset` method. So that it is easier to extend the states we need to temporarily bookkeep when coalescing is in flight.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

Differential Revision: [D71589949](https://our.internmc.facebook.com/intern/diff/D71589949)

[ghstack-poisoned]

kwen2501 commented:
@kwen2501 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71589949).

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -3325,8 +3312,15 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::collective(
   if (asyncOp) {
-    work->stashTensors(inputs);
-    work->stashTensors(outputs);
+    if (coalescingState_.active) {
+      coalescingState_.tensors.insert(
+          coalescingState_.tensors.end(), inputs.begin(), inputs.end());
+      coalescingState_.tensors.insert(
+          coalescingState_.tensors.end(), outputs.begin(), outputs.end());
+    } else {
+      work->stashTensors(inputs);
+      work->stashTensors(outputs);
+    }
```
This is the real fix to Torchrec's failure. It is using `reduce_scatter_v` and `all_gather_v`

kwen2501 commented:
Landed
